src/wallet/sendTransaction.ts

```diff
@@ -0,0 +1,353 @@
+      this.emit('estimation-updated', data);
+    })
+
+    this.mineTransaction.on('job-submitted', (data) => {
+      this.emit('job-submitted', data);
+    })
+
+    this.mineTransaction.on('job-done', (data) => {
+      this.emit('job-done', data);
+    })
+
+    this.mineTransaction.on('error', (message) => {
+      this.emit('send-error', message);
+    })
+
+    this.mineTransaction.on('unexpected-error', (message) => {
+      this.emit('send-error', message);
+    })
+
+    this.mineTransaction.on('success', (data) => {
+      this.transaction!.parents = data.parents;
+      this.transaction!.timestamp = data.timestamp;
+      this.transaction!.nonce = data.nonce;
+      this.transaction!.weight = data.weight;
+      this.emit('mine-tx-ended', data);
+    })
+
+    if (newOptions.startMiningTx) {
+      this.mineTransaction.start();
+    }
+
+    return this.mineTransaction;
+  }
+
+  /**
+   * Create and send a tx proposal to wallet service
+   * Expects this.transaction to be prepared, signed and mined
+   *
+   * @memberof SendTransactionWalletService
+   * @inner
+   */
+  async handleSendTxProposal() {
+    if (this.transaction === null) {
+      throw new WalletError('Can\'t push transaction if it\'s null.');
+    }
+    this.emit('send-tx-start', this.transaction);
+    const txHex = this.transaction.toHex();
+
+    try {
+      const responseData = await walletApi.createTxProposal(this.wallet, txHex);
+      const txProposalId = responseData.txProposalId;
+      const sendData = await walletApi.updateTxProposal(this.wallet, txProposalId, txHex);
+      this.transaction.updateHash();
+      this.emit('send-tx-success', this.transaction);
+    } catch (err) {
+      if (err instanceof WalletRequestError) {
+        this.emit('send-error', 'Error sending tx proposal.');
+      } else {
+        throw err;
+      }
+    }
+  }
+
+  /**
+   * Run sendTransaction from mining, i.e. expects this.transaction to be prepared and signed
+   * then it will mine and handle tx proposal
+   *
+   * 'until' parameter can be 'mine-tx', in order to only mine the transaction without propagating
+   *
+   * @memberof SendTransactionWalletService
+   * @inner
+   */
+  runFromMining(until = null) {
+    try {
+      this.mineTx();
+      if (until === 'mine-tx') {
+        return;
+      }
+
+      this.once('mine-tx-ended', (data) => {
+        this.handleSendTxProposal();
+      });
+    } catch (err) {
+      if (err instanceof WalletError) {
+        this.emit('send-error', err);
+      } else {
+        throw err;
+      }
+    }
+  }
+
+  /**
+   * Run sendTransaction from preparing, i.e. prepare, sign, mine and send the tx
+   *
+   * 'until' parameter can be 'prepare-tx' (it will stop before signing the tx), 'sign-tx' (it will stop before mining the tx),
+   * or 'mine-tx' (it will stop before send tx proposal, i.e. propagating the tx)
+   *
+   * @memberof SendTransactionWalletService
+   * @inner
+   */
+  async run(until = null) {
```

Comment:
I feel it could return a promise that will be resolved or rejected accordingly. What do you think?

Reply:
This is exactly what's being done in the wallet send transaction method.

```
  async sendManyOutputsTransaction(outputs: OutputRequestObj[], options: { inputs?: InputRequestObj[], changeAddress?: string } = {}): Promise<Transaction | string> {
    this.failIfWalletNotReady();
    const newOptions = Object.assign({
      inputs: [],
      changeAddress: null
    }, options);
    const { inputs, changeAddress } = newOptions;
    const sendTransaction = new SendTransactionWalletService(this, { outputs, inputs, changeAddress });
    const promise: Promise<Transaction | string> = new Promise((resolve, reject) => {
      sendTransaction.on('send-tx-success', (transaction) => {
        resolve(transaction);
      });

      sendTransaction.on('send-error', (err) => {
        reject(err);
      });
    });

    sendTransaction.run();
    return promise;
  }
```

At first I thought about having the sendTransaction only with the events. Should we add the promise also here?

Reply:
I feel it could return the promise and it would be used by `sendManyOutputsTransaction()` and others. I mean, the user can decide what's the best approach for their case and use the promise or the events accordingly.

Reply:
Yeah, I think I agree with you. Refactor done here https://github.com/HathorNetwork/hathor-wallet-lib/pull/221/commits/43cdd56eebf3e443132890aaddd495d24aaf24f5

Reply:
That's great! Thanks!